libimage/filters.go

```diff
@@ -265,62 +281,72 @@ func filterManifest(ctx context.Context, value bool) filterFunc {
 
 // filterReferences creates a reference filter for matching the specified wantedReferenceMatches value (OR logic)
 // and for matching the unwantedReferenceMatches values (AND logic)
-func filterReferences(r *Runtime, wantedReferenceMatches, unwantedReferenceMatches []string) filterFunc {
-	return func(img *Image, _ *layerTree) (bool, error) {
+func filterReferences(r *Runtime, wantedReferenceMatches, unwantedReferenceMatches []string) referenceFilterFunc {
+	return func(img *Image) (bool, []string, error) {
+		namesThatMatch := img.Names()
 		// Empty reference filters, return true
 		if len(wantedReferenceMatches) == 0 && len(unwantedReferenceMatches) == 0 {
-			return true, nil
+			return true, namesThatMatch, nil
 		}
 
 		unwantedMatched := false
+		// TODO 6.0 podman: remove unwanted matches from the output names. https://github.com/containers/common/pull/2413#discussion_r2031749013
 		// Go through the unwanted matches first
 		for _, value := range unwantedReferenceMatches {
-			matches, err := imageMatchesReferenceFilter(r, img, value)
+			matchedNames, err := imageMatchesReferenceFilter(r, img, value)
 			if err != nil {
-				return false, err
+				return false, namesThatMatch, err
 			}
-			if matches {
+			if len(matchedNames) > 0 {
 				unwantedMatched = true
 			}
 		}
 
 		// If there are no wanted match filters, then return false for the image
 		// that matched the unwanted value otherwise return true
 		if len(wantedReferenceMatches) == 0 {
-			return !unwantedMatched, nil
+			return !unwantedMatched, namesThatMatch, nil
 		}
 
 		// Go through the wanted matches
 		// If an image matches the wanted filter but it also matches the unwanted
 		// filter, don't add it to the output
 		for _, value := range wantedReferenceMatches {
-			matches, err := imageMatchesReferenceFilter(r, img, value)
+			matchedNames, err := imageMatchesReferenceFilter(r, img, value)
 			if err != nil {
-				return false, err
+				return false, nil, err
 			}
-			if matches && !unwantedMatched {
-				return true, nil
+			if len(matchedNames) > 0 && !unwantedMatched {
+				// Removes non-compliant names from image names
+				namesThatMatch = slices.DeleteFunc(namesThatMatch, func(name string) bool {
+					return !slices.ContainsFunc(matchedNames, func(matchedName string) bool {
+						return strings.Contains(name, matchedName)
+					})
+				})
+				return true, namesThatMatch, nil
```

Comment:
- `strings.Contains` is imprecise to the point of incorrectness
- How does this work for multi-element `wantedReferenceMatches`? AFAICS this returns incorrect values, only for the first element, in that case